File: rude_chairs.js
//modified linked list tutorial in javascript: http://www.i-programmer.info/programming/javascript/5328-javascript-data-structures-the-linked-list.html

function List(){
	this.start=null;
	this.end=null;
};

//creates the node in the double linked list (eg., it has reference points to the previous and next node in the list)
List.prototype.makeNode = function(){
	return {data: null, next: null} //prev: null, index: null};
}

List.prototype.add = function(data){

	var i = 1;

	if(this.start==null){
		//this is a new list, make a start node, then link it to the end node
		this.start=this.makeNode();
		this.end=this.start;
	}
	else{
		//this list exists, point the end of the list to the new node
		this.end.next=this.makeNode();
		this.end=this.end.next;
	}
	
	this.end.data = data;
};

List.prototype.circular = Function(){

	//find the end node, link the next node to the start node.
	this.end.next = this.start;

};

List.prototype.addAtIndex = function(index){};

List.prototype.removeByIndex = function(index){};


//find a certain node at D data value
List.prototype.findData = function(data){
	
	var currentNode = this.start;

	while (currentNode.data !== data){
		currentNode = this.getNext(currentNode);
	}

	return currentNode;
};

//find a certain node at N index value
List.prototype.findIndex = function(index){
	
	var currentNode = this.start;

	while (currentNode.index !== index){
		currentNode = this.getNext(currentNode);
	}
	return currentNode;
};

List.prototype.count = function(){
	var n = 1;
	var currentNode = this.start;

 	while (currentNode.next !== null){
		//console.log(current.next);
		n++
		currentNode = currentNode.next;
	}
	return n;
};

//traverse the list by 1 node
List.prototype.getNext = function(currentNode){

 	if (currentNode.next !== null){
		currentNode = currentNode.next;
		return currentNode;
	}
	else {
		//End of the line
		return null;
	}
};

List.prototype.print = function(){

	var currentNode = this.start;
	console.log("--- Start Of List ---");
	while (currentNode.next !== null){
		console.log('------------------------');
		console.log('List Node Index: '+currentNode.index);
		console.log('List Node Data: '+currentNode.data);
		currentNode = this.getNext(currentNode);
	}
	return "--- End Of List ---";
};

//insert a node at the start of the list
List.prototype.prepend = function(data){
	var tempNode = this.makeNode();
	tempNode.next = this.start;
	this.start = tempNode;
	tempNode.data = data;
};


function rudeChairs(numOfChairs, numOfPeople){

	var list = new List();

	for (var i = 1; i <= numOfChairs; i++){
		list.add(i);
	}

	var emptyChairs = 0;
	var currentNode = list.start;

	return console.log(list);
}

rudeChairs(100,100);

File: rude_chairs.py
#You are in a room with a circle of 100 chairs. The chairs are numbered sequentially from 1 to 100.
#At some point in time, the person in chair #1 will be asked to leave. The person in chair #2 will be skipped, and the person in chair #3 will be asked to leave. This pattern of skipping one person and asking the next to leave will keep going around the circle until there is one person left the survivor.
#Write a program to determine which chair the survivor is sitting in.

class Node:
	def __init__(self, initData):
		self.data = initData
		self.next = None

	def getData(self):
		return self.data

	def getNext(self):
		return self.next

class LinkedList:
	def __init__(self):
		self.head = None
		self.tail = None

	def append(self, item):
		tempNode = Node(item)
		
		if self.head:

			#makes the linked list circular			
			tempNode.next = self.head
			
			if self.tail:
				self.tail.next = tempNode
				self.tail = tempNode
			else:
				self.tail = tempNode
				self.head.next = self.tail

		else:
			self.head = tempNode

	def printList(self):

		currentNode = self.head

		for i in range(self.count()):
			print 'Current Node Data:',currentNode.data
			currentNode = currentNode.next
		return "end of list"

	def count(self):

		counter = 1
		currentNode = self.head

		while currentNode.next != self.head:
			counter +=1
			currentNode = currentNode.next
		return counter

	#Note: returns the node object
	def find(self, data):
		currentNode = self.head

		for i in range(self.count()):
			if currentNode.data != data:
				currentNode = currentNode.next
			else:
				return currentNode
	
	def delete(self, data):

		#find the node
		removeNode = self.find(data)

		#find the previous node	
		refNode = self.head

		for i in range(self.count()):
			if removeNode != refNode.next:
				refNode = refNode.next
		
		#swap pointers
		if removeNode == self.head:
			self.head = removeNode.next
			refNode.next = self.head

		elif removeNode == self.tail:
			self.tail = removeNode.next
			refNode.next = self.tail
		
		else:
			refNode.next = removeNode.next


def rude_chairs(num_of_chairs):

	chairs = LinkedList()

	for num in range(1, num_of_chairs+1):
		chairs.append(num)

	currentNode = chairs.head
	deleteNode = None

	while chairs.count() > 1:

		deleteNode = currentNode

		currentNode = currentNode.next.next

		#print "Delete Chair: ", deleteNode.data

		chairs.delete(deleteNode.data)

	return 'Survivor is', chairs.head.data

print rude_chairs(10)